tests/unit/s2n_connection_test.c

```diff
@@ -522,6 +524,83 @@ int main(int argc, char **argv)
         EXPECT_SUCCESS(s2n_connection_free(conn));
     }
 
+    /* s2n_connection_set_fd can be called twice in a row */
+    {
+        static const int OLDFD = 1;
+        static const int NEWFD = 2;
+        static int getReadFd, getWriteFd;
+
+        struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT);
+        EXPECT_NOT_NULL(conn);
+
+        EXPECT_SUCCESS(s2n_connection_set_fd(conn, OLDFD));
+        EXPECT_SUCCESS(s2n_connection_set_fd(conn, NEWFD));
+
+        EXPECT_SUCCESS(s2n_connection_get_write_fd(conn, &getWriteFd));
+        EXPECT_SUCCESS(s2n_connection_get_read_fd(conn, &getReadFd));
+        EXPECT_EQUAL(getReadFd, NEWFD);
+        EXPECT_EQUAL(getWriteFd, NEWFD);
+
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+    }
+
+    /* s2n_connection_set_read_fd and s2n_connection_set_write_fd can be called 
+     * after s2n_connection_set_fd */
+    {
+        static const int OLDFD = 1;
+        static const int NEWFD = 2;
+        static int getReadFd, getWriteFd;
+
+        struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT);
+        EXPECT_NOT_NULL(conn);
+
+        EXPECT_SUCCESS(s2n_connection_set_fd(conn, OLDFD));
+        EXPECT_SUCCESS(s2n_connection_set_read_fd(conn, NEWFD));
+        EXPECT_SUCCESS(s2n_connection_set_write_fd(conn, NEWFD));
+
+        EXPECT_SUCCESS(s2n_connection_get_write_fd(conn, &getWriteFd));
+        EXPECT_SUCCESS(s2n_connection_get_read_fd(conn, &getReadFd));
+        EXPECT_EQUAL(getReadFd, NEWFD);
+        EXPECT_EQUAL(getWriteFd, NEWFD);
+
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+    }
+
+    /* The default s2n socket read/write setup can be used with a user-defined send/recv context */
+    {
```

Comment:
A customer doesn't have access to the internal structure we expect for the default socket read/write, so there's no way they could set a valid custom context to use with that callback. We need the managed/custom status of the callback and context to match.

Maybe set the callback to NULL if the context is set first and io was managed (and vice versa)?